### ISSUE TITLE
API discussion

### ISSUE THREAD

**Issue by 19h** [sysdiglabs/sysdig-sdk-python#5]
Hey,

lets discuss things here :-)

Things that I mentioned at other places:
- I'd like to have continuous integration testing of the API (I'll do this with my client, at least. Not sure about the API rate limit though?)
- Inconsistent gzipping of responses (get_views, et al.)

**Comment by KenanSulayman** [comment]
The API should be a bit streamlined. Some endpoints return data directly, others wrap the payload.

E.g.:

`/ui/dashboards` is returning `{"dashboards":[...]}`
`/api/data/metrics` is returning `{...}`

I guess the issue is that your contract is to return a hash map?

**Comment by gianlucaborello** [comment]
Thanks for consolidating the issues here. To address your points:

- Integration testing is certainly fine, and as long as it's done reasonably (e.g. up to a few requests per second) you shouldn't incur in any rate limit. For our Python wrapper we currently have some automatic tests, albeit very rudimentary. Essentially, every time we add a new wrapper method we try to add an example that uses it (so our customers can have an almost ready snippet of code to use), and then use that example in our tests https://github.com/draios/python-sdc-client/blob/master/.travis.yml#L9 Certainly dedicated integration tests in addition to the examples would be better, we just haven't found the time to go ahead with that.

- Inconsistent gzipping: good catch, all our API is served by AWS Cloudfront and some static assets (like the json you were reporting) are hosted in S3 and manually gzipped. This is because we started using Cloudfront before automatic gzip support was released at the end of last year (https://aws.amazon.com/blogs/aws/new-gzip-compression-support-for-amazon-cloudfront/). We should definitely change the deployment of our static assets and let Cloudfront serve both, now that it supports that.

- API format: you are right, the API is not super consistent. We tend to adopt a model where we always wrap the resource in a separate JSON object. I will check with the frontend developers, but at the moment we can't hardcode something like `{ 'data' : {} }` because I believe our frontend framework expects the key to be somewhat meaningful, and our frontend at the moment is by far the most frequent client of this API :) What we could certainly do, and would be less invasive for us at the moment, would be to make sure that the various API clients such as our Python one automatically wrap and unwrap the resources so the client doesn't have to worry about them (this is done in some methods, but again not super consistent all over the place).

**Comment by KenanSulayman** [comment]
Great, I'll update you on the testing part once I have implemented the parameterized API calls. (Progress: https://github.com/KenanSulayman/sysdig-node/issues/1)

The programmatic unwrapping isn't a problem from a mapping perspective and I have inherited that from the Python client ([sysdig-client.js#L134](https://github.com/KenanSulayman/sysdig-node/blob/master/lib/sysdig-client.js#L134)). I was just talking about the API when working with it directly, though some operations are definitely too logic-heavy to be handled by a sane client (e.g. get_notification_ids or add_dashboard_panel, ...). It's not at all a priority right now, but at least having some kind of grace period and communication strategy for API changes would be 😲.
